File: example02threading/src/androidTest/java/foo/bar/example/forethreading/ui/StateBuilder.java
package foo.bar.example.forethreading.ui;

import androidx.test.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;

import foo.bar.example.forethreading.CustomApp;
import foo.bar.example.forethreading.feature.counter.CounterWithLambdas;
import foo.bar.example.forethreading.feature.counter.CounterWithProgress;

import static org.mockito.Mockito.when;

/**
 *
 */
public class StateBuilder {

    private CounterWithLambdas mockCounterWithLambdas;
    private CounterWithProgress mockCounterWithProgress;

    StateBuilder(CounterWithLambdas mockCounterWithLambdas, CounterWithProgress mockCounterWithProgress) {
        this.mockCounterWithLambdas = mockCounterWithLambdas;
        this.mockCounterWithProgress = mockCounterWithProgress;
    }

    StateBuilder counterBasicIsBusy(boolean busy) {
        when(mockCounterWithLambdas.isBusy()).thenReturn(busy);
        return this;
    }

    StateBuilder counterWithProgressIsBusy(boolean busy) {
        when(mockCounterWithProgress.isBusy()).thenReturn(busy);
        return this;
    }

    StateBuilder counterBasicCount(int count) {
        when(mockCounterWithLambdas.getCount()).thenReturn(count);
        return this;
    }

    StateBuilder counterWithProgressCount(int count) {
        when(mockCounterWithProgress.getCount()).thenReturn(count);
        return this;
    }

    StateBuilder counterWithProgressProgressValue(int value) {
        when(mockCounterWithProgress.getProgress()).thenReturn(value);
        return this;
    }

    ActivityTestRule<CounterActivity>  createRule(){

        return new ActivityTestRule<CounterActivity>(CounterActivity.class) {
            @Override
            protected void beforeActivityLaunched() {

                //get hold of the application
                CustomApp customApp = (CustomApp) InstrumentationRegistry.getTargetContext().getApplicationContext();
                customApp.injectSynchronousObjectGraph();

                //inject our mocks so our UI layer will pick them up
                customApp.injectMockObject(CounterWithLambdas.class, mockCounterWithLambdas);
                customApp.injectMockObject(CounterWithProgress.class, mockCounterWithProgress);
            }

        };
    }

}

File: example-kt-02coroutine/src/androidTest/java/foo/bar/example/forecoroutine/ui/StateBuilder.kt
package foo.bar.example.forecoroutine.ui

import androidx.test.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import foo.bar.example.forecoroutine.App
import foo.bar.example.forecoroutine.OG
import foo.bar.example.forecoroutine.feature.counter.Counter
import foo.bar.example.forecoroutine.feature.counter.CounterWithProgress
import io.mockk.every


/**
 * Copyright © 2019 early.co. All rights reserved.
 */
class StateBuilder internal constructor(private val mockCounter: Counter, private val mockCounterWithProgress: CounterWithProgress) {

    internal fun counterBasicIsBusy(busy: Boolean): StateBuilder {
        every {
            mockCounter.isBusy
        } returns busy
        return this
    }

    internal fun counterWithProgressIsBusy(busy: Boolean): StateBuilder {
        every {
            mockCounterWithProgress.isBusy
        } returns busy
        return this
    }

    internal fun counterBasicCount(count: Int): StateBuilder {
        every {
            mockCounter.count
        } returns count
        return this
    }

    internal fun counterWithProgressCount(count: Int): StateBuilder {
        every {
            mockCounterWithProgress.count
        } returns count
        return this
    }

    internal fun counterWithProgressProgressValue(value: Int): StateBuilder {
        every {
            mockCounterWithProgress.progress
        } returns value
        return this
    }

    internal fun createRule(): ActivityTestRule<CounterActivity> {

        return object : ActivityTestRule<CounterActivity>(CounterActivity::class.java) {
            override fun beforeActivityLaunched() {

                //get hold of the application
                val customApp = InstrumentationRegistry.getTargetContext().applicationContext as App
                // customApp.injectSynchronousObjectGraph()

                //inject our mocks so our UI layer will pick them up
                OG.putMock(Counter::class.java, mockCounter)
                OG.putMock(CounterWithProgress::class.java, mockCounterWithProgress)
            }

        }
    }

}
